### ISSUE TITLE
Starcounter.DbException - ScErrReadOnlyTransaction

### ISSUE THREAD

**Issue by PatrykSzwer** [StarcounterApps/Website#176]
### Steps to reproduce
1. Reset data.
2. Click on DefaultSurface
(If error didn't appear, go back to SurfaceGroups and click DefaultSurface again.)
### Expected result
No errors.
### Actual result
Error with connection lost is showing up:
![image](https://user-images.githubusercontent.com/25042256/29317587-f0d3002c-81cc-11e7-96b7-534b0f6cc197.png)

##### Starcounter version 2.3.1.6924

As We already discussed with @joozek78 It can be related to authorization library and DbScopeMiddleware, but as It wasn't discovered before and It's pretty easy to find bug, It can also be related to SC version. Needs further investigation. 

cc @un-tone

**Comment by un-tone** [comment]
Part of discussion is here https://github.com/StarcounterApps/Website/pull/178. I still have not found a reason of the issue. I already don't have time to do it. But I found out that this https://github.com/StarcounterApps/Website/pull/161 implementation works well without that bug. So I have to ask @chrhol to escalate that #161 issue.

**Comment by un-tone** [comment]
After changes in SC done for https://github.com/StarcounterApps/Website/issues/177 the behaviour was changed a little. But it is still reproducible if you will got catching rules of SidebarSurface and uncheck "Nestable" checkbox for the rule.

**Comment by un-tone** [comment]
I checked the issue in #161 with last changes. Unfortunately that changes does not fix the issue, maybe I missed something last time.

**Comment by un-tone** [comment]
### Steps to reproduce

1. Run **WebsiteEditor** and **WebsiteProvider** apps from `develop` branch
1. Run **SignIn** app from `develop` branch
1. If a user is not created go to <http://localhost:8080/signin/generateadminuser>
1. Go to <http://localhost:8080/signin/signinuser> and log in
1. Go to <http://localhost:8080/websiteeditor>
1. Go to `SidebarSurface` page (click on the link)

### Expected result

The page is loaded without errors.

### Actual result

Get the error:

```
Starcounter.DbException: ScErrReadOnlyTransaction (SCERR4093): The transaction is readonly and cannot be changed to write-mode.
Version: 2.3.1.7063.
Help page: https://github.com/Starcounter/Starcounter/wiki/SCERR4093.
   в Starcounter.Internal.DbState.WriteObjectRaw(UInt64 oid, UInt64 address, Int32 index, ObjectRef valueRef) в C:\TeamCity\BuildAgent\work\sc-7638\Level1\src\Starcounter\Internal\DbState.cs:строка 1151
   в Starcounter.Internal.DbState.WriteObject(UInt64 oid, UInt64 address, Int32 index, IObjectProxy value) в C:\TeamCity\BuildAgent\work\sc-7638\Level1\src\Starcounter\Internal\DbState.cs:строка 1137
   в Simplified.Ring6.WebTemplate.set_<WebTemplateGroup>k__BackingField(WebTemplateGroup value)
   в WebsiteEditor.ViewModels.GeneralPage.Handle(CurrentGroupKey action) в C:\Users\untone\Documents\Projects\Starcounter\Git\StarcounterApps\Website\src\WebsiteEditor\ViewModels\GeneralPage.json.cs:строка 49
   в WebsiteEditor.ViewModels.GeneralPage.JsonByExample.Schema.<>c.<.ctor>b__0_15(Json pup, Input`1 input)
   в Starcounter.Templates.Property`1.InvokeHandler(Json parent, Input`1 input) в C:\TeamCity\BuildAgent\work\sc-7638\Level1\src\Starcounter.XSON\Templates\Foundation\Property.cs:строка 267
   в Starcounter.Templates.Property`1.ProcessInput(Json parent, T value) в C:\TeamCity\BuildAgent\work\sc-7638\Level1\src\Starcounter.XSON\Templates\Foundation\Property.cs:строка 289
   в Starcounter.XSON.DefaultPatchHandler.<>c__DisplayClass0_0.<Handle>b__0() в C:\TeamCity\BuildAgent\work\sc-7638\Level1\src\Starcounter.XSON\Sessions\JSON-Patch\DefaultPatchHandler.cs:строка 50
   в Starcounter.Advanced.XSON.JsonExtension.Scope(Json json, Action action) в C:\TeamCity\BuildAgent\work\sc-7638\Level1\src\Starcounter.XSON\Advanced\JsonExtension.cs:строка 69
   в Starcounter.XSON.DefaultPatchHandler.Handle(Json root, JsonPatchOperation patchOp, JsonPointer pointer, String value) в C:\TeamCity\BuildAgent\work\sc-7638\Level1\src\Starcounter.XSON\Sessions\JSON-Patch\DefaultPatchHandler.cs:строка 38
   в Starcounter.XSON.JsonPatch.HandleOnePatch(Json root, JsonPatchOperation op, JsonPointer ptr, String value, Int32 patchCount, ChangeLog changeLog, Object sourceJson, Boolean strictRejection) в C:\TeamCity\BuildAgent\work\sc-7638\Level1\src\Starcounter.XSON\Sessions\JSON-Patch\JsonPatch.cs:строка 546
   в Starcounter.XSON.JsonPatch.Apply(Json root, JsonTextReader reader, Boolean strictPatchRejection, Object sourceJson, Int32& patchCount) в C:\TeamCity\BuildAgent\work\sc-7638\Level1\src\Starcounter.XSON\Sessions\JSON-Patch\JsonPatch.cs:строка 454
   в Starcounter.XSON.JsonPatch.Apply(Json root, String patch, Boolean strictPatchRejection, Int32& patchCount) в C:\TeamCity\BuildAgent\work\sc-7638\Level1\src\Starcounter.XSON\Sessions\JSON-Patch\JsonPatch.cs:строка 376
   в Starcounter.Internal.PuppetRestHandler.HandleWebSocketJson(String bs, WebSocket ws) в C:\TeamCity\BuildAgent\work\sc-7638\Level1\src\Starcounter.Apps.JsonPatch\BuiltInRestHandlers\PuppetRestHandler.cs:строка 62
   в Starcounter.Internal.AllWsGroups.RunHandler(UInt16 managedHandlerId, UInt32 groupId, WebSocket ws) в C:\TeamCity\BuildAgent\work\sc-7638\Level1\src\Starcounter.Internal\Http\WsManager.cs:строка 289
   в Starcounter.GatewayHandlers.HandleWebSocket(UInt16 managedHandlerId, Byte* rawChunk, BMX_TASK_INFO* taskInfo, Boolean* isHandled) в C:\TeamCity\BuildAgent\work\sc-7638\Level1\src\Starcounter\Apps\GatewayHandlers.cs:строка 793
ErrorCode=4093
HResult=-2146233088
HelpLink=https://github.com/Starcounter/Starcounter/wiki/SCERR4093
```
### Technical information

##### Starcounter version 2.3.1.7063

### Screenshot
![image](https://user-images.githubusercontent.com/1625378/29725484-f8cb9496-89d4-11e7-9b27-216755235209.png)

**Comment by un-tone** [comment]
I refreshed the steps above. I can't catch the reason why in this case it throws such exception. I already have no idea how to investigate it further.
@chrhol could you please take a look at this case and get me any clues? Thanks in advance.

**Comment by chrhol** [comment]
I started looking into this but did not have enough time to finish. 

I have to admit that trying to debug the code presented me with some problems with different middlewares making each step kind of obfuscated.

It's possible I found something that doesn't work as intended but I'm not sure yet. Need to continue tomorrow

**Comment by chrhol** [comment]
I think I found the problem (finally). The attached transaction is set to null in a certain place when a viewmodel is removed (internally) from what the session keeps track of, and then at a later stage the same viewmodel is readded, but of course the transaction is gone. 

I am not really sure why the transaction is set to null. Commenting out that line seems to work fine, but there should be some reason for it. I will have to try to find why it was added before doing a fix.

**Comment by chrhol** [comment]
So apparently there was a fix for a crash (https://github.com/Starcounter/level1/issues/3525 for those that still can see level1). Not sure how to solve this yet.

**Comment by PatrykSzwer** [comment]
Any updates with this?

**Comment by un-tone** [comment]
The issue was partially fixed in last SC build 2.3.1.7119. Thank you @chrhol!
Currently it doesn't throw the exception but it doesn't refresh page in the described case when you go from general surface page to catching rule page (or any other surface page). I'll try to investigate it further.

**Comment by PatrykSzwer** [comment]
Since the main problem described in this issue is fixed, then I think we can close this issue and create a new one, regarding last @un-tone comment. I'll close this one for now.
